server2/Arcadia.Assistant/Arcadia.Assistant.ExternalStorages.SharepointOnline/SharepointAuthTokenService.cs

```diff
@@ -0,0 +1,214 @@
+
+            var requestData =
+                "grant_type=" + HttpUtility.UrlEncode(OAuth2GrantType) +
+                "&client_id=" + HttpUtility.UrlEncode(tokenClientId) +
+                "&client_secret=" + HttpUtility.UrlEncode(tokenClientSecret) +
+                "&resource=" + HttpUtility.UrlEncode(tokenResource);
+
+            var request = new HttpRequestMessage(HttpMethod.Post, tokenUrl)
+            {
+                Content = new StringContent(requestData)
+            };
+
+            request.Content.Headers.Remove("Content-Type");
+            request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/x-www-form-urlencoded");
+            request.Content.Headers.ContentLength = requestData.Length;
+
+            var response = await this.httpClient.SendAsync(request, cancellationToken);
+
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new HttpRequestException("Access token response has failed");
+            }
+
+            var responseContent = await response.Content.ReadAsStringAsync();
+
+            var oauthResult = JsonSerializer.Deserialize<OAuth2AccessTokenResponse>(responseContent);
+            return oauthResult.AccessToken;
+        }
+
+        private async Task<string?> GetOAuth2Url(string realm, CancellationToken cancellationToken)
+        {
+            var metadata = await this.GetMetadataDocument(realm, cancellationToken);
+
+            var s2SEndpoint = metadata.Endpoints?.SingleOrDefault(e => e.Protocol == OAuth2Protocol);
+
+            if (s2SEndpoint != null)
+            {
+                return s2SEndpoint.Location;
+            }
+
+            throw new Exception("Metadata document does not contain OAuth2 endpoint url");
+        }
+
+        private async Task<JsonMetadataDocument> GetMetadataDocument(string realm, CancellationToken cancellationToken)
+        {
+            var metadataEndpointUrl = $"{AcsMetadataEndpoint}?realm={realm}";
+
+            var response = await this.httpClient.GetAsync(metadataEndpointUrl, cancellationToken);
+            var metadataString = await response.Content.ReadAsStringAsync();
+            var metadata = JsonSerializer.Deserialize<JsonMetadataDocument>(metadataString);
+
+            if (metadata == null)
+            {
+                throw new Exception($"No metadata document found at the global endpoint {metadataEndpointUrl}");
+            }
+
+            return metadata;
+        }
+
+        private string GetFormattedPrincipal(string principalName, string? hostName, string realm)
+        {
+            return string.IsNullOrEmpty(hostName)
+                ? string.Format(CultureInfo.InvariantCulture, "{0}@{1}", principalName, realm)
+                : string.Format(CultureInfo.InvariantCulture, "{0}/{1}@{2}", principalName, hostName, realm);
+        }
+
+        [DataContract]
+        private class JsonMetadataDocument
+        {
+            [JsonPropertyName("endpoints")]
```

Comment:
Why is this needed?

Reply:
Ok, probably I've found the source of evil.
https://github.com/dotnet/runtime/issues/29975
So then, could you use `JsonPropertyName` instead `DataMember` in every single case without mixing of them?

Reply:
Or we can try to use `Newtonsoft.JSON` instead of standard Json serializer, it respects `DataMember` attributes. @santee?

Reply:
We replace Newtonsoft lib by MS lib Text.Json by @santee requirement. I will remove redundant Data serialization attributes.